module/zfs/dsl_scan.c

```diff
@@ -2038,91 +3467,427 @@ dsl_scan_scrub_cb(dsl_pool_t *dp,
+	 *	extent boundaries.
+	 * 2) Hot, where the zio is currently in-flight in
+	 *	dsl_scan_issue_ios. In this case, we can't simply
+	 *	reach in and stop the in-flight zio's, so we instead
+	 *	block the caller. Eventually, dsl_scan_issue_ios will
+	 *	be done with issuing the zio's it gathered and will
+	 *	signal us.
 	 */
-	spa_vdev_state_enter(spa, SCL_NONE);
-	spa->spa_scrub_reopen = B_TRUE;
-	vdev_reopen(spa->spa_root_vdev);
-	spa->spa_scrub_reopen = B_FALSE;
-	(void) spa_vdev_state_exit(spa, NULL, 0);
+	sio = avl_find(&queue->q_zios_by_addr, &srch, &idx);
+	if (sio != NULL) {
+		range_seg_t *rs;
+		int64_t asize = sio->sio_asize;
+		blkptr_t tmpbp;
 
-	if (func == POOL_SCAN_SCRUB && dsl_scan_is_paused_scrub(scn)) {
-		/* got scrub start cmd, resume paused scrub */
-		int err = dsl_scrub_set_pause_resume(scn->scn_dp,
-		    POOL_SCRUB_NORMAL);
-		if (err == 0)
-			return (SET_ERROR(ECANCELED));
+		/* Got it while it was cold in the queue */
+		ASSERT3U(start, ==, SCAN_IO_GET_OFFSET(sio));
+		ASSERT3U(size, ==, asize);
+		avl_remove(&queue->q_zios_by_addr, sio);
 
-		return (SET_ERROR(err));
-	}
+		rs = range_tree_find(queue->q_exts_by_addr, start, size);
+		ASSERT(rs != NULL);
 
-	return (dsl_sync_task(spa_name(spa), dsl_scan_setup_check,
-	    dsl_scan_setup_sync, &func, 0, ZFS_SPACE_CHECK_NONE));
+		/*
+		 * Range trees with gaps do not support partial removes, so
+		 * instead we simply decrement the fill. The range tree code
+		 * will call the q_exts_by_size callbacks for us in either case.
+		 */
+		if (rs->rs_fill == size) {
+			range_tree_remove(queue->q_exts_by_addr,
+			    rs->rs_start, rs->rs_end - rs->rs_start);
+		} else {
+			range_tree_adjust_fill(queue->q_exts_by_addr,
+			    rs, -size);
+		}
+
+		/*
+		 * We only update scn_bytes_pending in the cold path,
+		 * otherwise it will already have been accounted for as
+		 * part of the zio's execution.
+		 */
+		atomic_add_64(&scn->scn_bytes_pending, -asize);
+
+		/* count the block as though we issued it */
+		sio2bp(sio, &tmpbp, dva_i);
+		count_block(scn, dp->dp_blkstats, &tmpbp);
+
+		kmem_free(sio, sizeof (*sio));
+	} else {
+		/*
+		 * If it's part of an extent that's currently being issued or
+		 * it isn't in the queue at all, wait until the extent has been
+		 * consumed. In this case we simply waited for the zio to
+		 * complete, so the acccounting will be done for us.
+		 */
+		while (queue->q_issuing_start <= start &&
+		    queue->q_issuing_end >= start + size) {
+			cv_wait(&queue->q_cv, &vdev->vdev_scan_io_queue_lock);
+		}
+	}
+	mutex_exit(q_lock);
 }
 
-static boolean_t
-dsl_scan_restarting(dsl_scan_t *scn, dmu_tx_t *tx)
+/*
+ * Callback invoked when a zio_free() zio is executing. This needs to be
+ * intercepted to prevent the zio from deallocating a particular portion
+ * of disk space and it then getting reallocated and written to, while we
+ * still have it queued up for processing, or even while we're trying to
+ * scrub or resilver it.
```

Comment:
This implies that we might leave the scrub/resilver i/os going after we return from dsl_scan_sync().  Is that right?